File: core/RandomPickWithBlacklist/RandomPickWithBlacklist.java
package com.dbc.code;

import java.util.HashMap;
import java.util.Map;
import java.util.Random;

public class RandomPickWithBlacklist {
    private final Map<Integer, Integer> map = new HashMap<>();
    private final Random random = new Random();
    private int length = 0;

    public RandomPickWithBlacklist(int n, int[] blacklist) {
        HashMap<Integer, Integer> remain = new HashMap<>();
        for (int black : blacklist) remain.put(black, 0);
        this.length = n - blacklist.length;
        for (int black : blacklist) {
            if (black < this.length) {
                while (remain.containsKey(n - 1)) n--;
                this.map.put(black, n - 1);
                n--;
            }
        }
    }

    public int pick() {
        int point = random.nextInt(this.length);
        return this.map.getOrDefault(point, point);
    }
}

File: core/RandomPickWithBlacklist/RandomPickWithBlacklist.py
import random
from typing import List

class Solution:
    def __init__(self, n: int, blacklist: List[int]):
        self.map = {}
        self.length = n - len(blacklist)
        for black in blacklist:
            if black < self.length:
                while n - 1 in blacklist:
                    n -= 1
                self.map[black] = n - 1
                n -= 1

    def pick(self) -> int:
        point = random.randint(0, self.length - 1)
        return point if point not in self.map else self.map[point]




